deepinv/sampling/samplers.py

```diff
@@ -0,0 +1,347 @@
+    :param float burnin_ratio: Percentage of iterations used for burn-in period (between 0 and 1). Default: 0.2
+    :param int thinning: Integer to thin the Monte Carlo samples (keeping one out of `thinning` samples). Default: 10
+    :param float thresh_conv: The convergence threshold for the mean and variance. Default: ``1e-3``
+    :param Callable callback: A funciton that is called on every (thinned) sample for diagnostics.
+    :param int history_size: Number of most recent samples to store in memory. Default: 5
+    :param bool verbose: Whether to print progress of the algorithm. Default: ``False``
+    """
+
+    def __init__(
+        self,
+        iterator: SamplingIterator,
+        data_fidelity: DataFidelity,
+        prior: Prior,
+        params_algo={"lambda": 1.0, "stepsize": 1.0},
+        max_iter=100,
+        callback=lambda x: x,
+        burnin_ratio=0.2,
+        thresh_conv=1e-3,
+        crit_conv="residual",
+        thinning=10,
+        history_size: Union[int, bool] = 5,
+        verbose=False,
+    ):
+        super(BaseSample, self).__init__()
+        self.iterator = iterator
+        self.data_fidelity = data_fidelity
+        self.prior = prior
+        self.params_algo = params_algo
+        self.max_iter = max_iter
+        self.burnin_ratio = burnin_ratio
+        self.thresh_conv = thresh_conv
+        self.crit_conv = crit_conv
+        self.callback = callback
+        self.mean_convergence = False
+        self.var_convergence = False
+        self.thinning = thinning
+        self.verbose = verbose
+        # self.clip = clip
+        self.history_size = history_size
+
+        # initialize history, to zero
+        if history_size is True:
+            self.history = []
+        elif history_size:
+            self.history = deque(maxlen=history_size)
+        else:
+            self.history = False
+
+    def forward(
+        self,
+        y: torch.Tensor,
+        physics: Physics,
+        X_init: Union[torch.Tensor, None] = None,
+        seed: Union[int, None] = None,
+    ) -> torch.Tensor:
+        r"""
+        Run the MCMC sampling chain and return the posterior sample mean.
+
+        :param torch.Tensor y: The observed measurements
+        :param Physics physics: Forward operator of your inverse problem
+        :param torch.Tensor X_init: Initial state of the Markov chain. If None, uses ``physics.A_adjoint(y)`` as the starting point
+            Default: ``None``
+        :param int seed: Optional random seed for reproducible sampling.
+            Default: ``None``
+        :return: Posterior sample mean
+        :rtype: torch.Tensor
+        """
+
+        # pass back out sample mean
+        return self.sample(y, physics, X_init=X_init, seed=seed)[0]
+
+    def sample(
+        self,
+        y: torch.Tensor,
+        physics: Physics,
+        X_init: Union[torch.Tensor, None] = None,
+        seed: Union[int, None] = None,
+        g_statistics=[lambda x: x],
```

Comment:
I'd be good to add the type here as well
```suggestion
        g_statistics: list[Callable[..., Any]] = [lambda x: x],
```